core/src/sumeragi/main_loop.rs

```diff
@@ -1138,7 +1179,13 @@ pub(crate) fn run(
 
             // NOTE: View change must be periodically suggested until it is accepted.
             // Must be initialized to pipeline time but can increase by chosen amount
-            view_change_time += state.world.view().parameters().sumeragi.pipeline_time();
+            view_change_time += state
+                .world
+                .view()
+                .parameters()
+                .sumeragi
+                .pipeline_time()
+                .saturating_mul((view_change_index + 1).try_into().unwrap_or(u32::MAX));
         }
 
         reset_state(
```

Comment:
Here `view_change_time` is nonlinear with `view_change_index`.
Also, doesn't this increment overlap with the one in the `reset_state()`?

Reply:
> Also, doesn't this increment overlap with the one in the reset_state()?

Inside `reset_state` `view_change_time` is reset to trigger view change index after `pipeline_time` * `view_change_index`.

> Here view_change_time is nonlinear with view_change_index.

Why not? `view_change_time` is increased in chunks of `pipeline_time` * `view_change_index` size.
So it would be like: 2, 4, 6, 8, ... so it's linear.

Reply:
[my branch]: https://github.com/s8sato/iroha/commit/54062b8fe5b5c1cffea59bc1d6818b739a94fad5

In [my branch] I found it was indeed linear when logged at line 981 but strange when logged at line 1189.

- [54062b8fe.log](https://github.com/user-attachments/files/16760619/54062b8fe.log)

This is part of my ERROR level log. I intentionally killed 2 peers to cause view changes

<details><summary>details</summary>

```txt
$ cat 54062b8fe.log | grep -B1 main_loop.rs:981

  2024-08-27T09:37:37.382184Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 1, view_change_time: 12s
    at core/src/sumeragi/main_loop.rs:981
--
  2024-08-27T09:37:49.410953Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 2, view_change_time: 18s
    at core/src/sumeragi/main_loop.rs:981
--
  2024-08-27T09:38:07.507803Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 3, view_change_time: 24s
    at core/src/sumeragi/main_loop.rs:981
--
  2024-08-27T09:38:31.535742Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 4, view_change_time: 30s
    at core/src/sumeragi/main_loop.rs:981
--
  2024-08-27T09:39:01.703077Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 5, view_change_time: 36s
    at core/src/sumeragi/main_loop.rs:981
```

```txt
$ cat 54062b8fe.log | grep -B1 main_loop.rs:1189

  2024-08-27T09:37:36.437753Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 18s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.451240Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 24s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.464787Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 30s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.478254Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 36s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.491821Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 42s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.505859Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 48s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.522229Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 54s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.536978Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 60s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.551447Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 66s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.565179Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 72s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.579937Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 78s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:36.593612Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 0, view_change_time: 84s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:37:49.395329Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 1, view_change_time: 24s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:38:07.419794Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 2, view_change_time: 36s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:38:31.520076Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 3, view_change_time: 48s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:39:01.547381Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 4, view_change_time: 60s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:39:37.704183Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 5, view_change_time: 72s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:40:13.707816Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 5, view_change_time: 108s
    at core/src/sumeragi/main_loop.rs:1189
--
  2024-08-27T09:40:49.703807Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 5, view_change_time: 144s
    at core/src/sumeragi/main_loop.rs:1189
```

</details>

Reply:
About line 1189, you can see that increments within the same view change index are constant. 
For:
- 0 it's 6 seconds
- 1 it's 12 seconds
- 2 it's 18 seconds
- 3 it's 24 seconds
- ...
So `view_change_time` is incremented with constant values within given `view_change_index`.
Sumeragi suggest a view change when since `last_view_change` elapsed more time than `view_change_time`.
Lets look for example at index 5:
- `adjusted_pipeline_time = pipeline_time * (view_change_index + 1) = 6s * 6 = 36s`
- so first view change would be suggested after 36 seconds since view change started:
```
  2024-08-27T09:39:01.703077Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 5, view_change_time: 36s
    at core/src/sumeragi/main_loop.rs:981
  2024-08-27T09:39:37.704183Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 5, view_change_time: 72s
    at core/src/sumeragi/main_loop.rs:1189
```
- you can see that ~36 seconds elapsed since between this 2 records
- the next attempt would be after 72 seconds after view change started (or 36 seconds after previous attempt):
```
  2024-08-27T09:39:37.704183Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 5, view_change_time: 72s
    at core/src/sumeragi/main_loop.rs:1189

  2024-08-27T09:40:13.707816Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 5, view_change_time: 108s
    at core/src/sumeragi/main_loop.rs:1189

  2024-08-27T09:40:49.703807Z ERROR iroha_core::sumeragi::main_loop: view_change_index: 5, view_change_time: 144s
    at core/src/sumeragi/main_loop.rs:1189
```
- again you can see that since first and the last record about 72 seconds elapsed (or 36 between last and previous)

Reply:
I see, so increments per index and increments per attempt within the same index are different things, and in the latter case `view_change_time` indicates not a duration between attempts but a point in time within the same index because the origin (`last_view_change_time`) is not reset